impl/src/main/java/io/kroxylicious/testing/kafka/common/Utils.java

```diff
@@ -53,28 +58,54 @@ public static Stream<ServerSocket> preAllocateListeningSockets(int num) {
     }
 
     public static void awaitExpectedBrokerCountInCluster(Map<String, Object> connectionConfig, int timeout, TimeUnit timeUnit, Integer expectedBrokerCount) {
-        try (Admin admin = Admin.create(connectionConfig)) {
-            Awaitility.await()
-                    .pollDelay(Duration.ZERO)
-                    .pollInterval(1, TimeUnit.SECONDS)
-                    .atMost(timeout, timeUnit)
-                    .ignoreExceptions()
-                    .until(() -> {
-                        log.info("describing cluster: {}", connectionConfig.get("bootstrap.servers"));
-                        final Collection<Node> nodes;
-                        try {
-                            nodes = admin.describeCluster().nodes().get(10, TimeUnit.SECONDS);
-                            log.info("got nodes: {}", nodes);
-                            return nodes;
-                        }
-                        catch (InterruptedException | ExecutionException e) {
-                            log.warn("caught: {}", e.getMessage(), e);
-                        }
-                        catch (TimeoutException te) {
-                            log.warn("Kafka timed out describing the the cluster");
-                        }
-                        return Collections.emptyList();
-                    }, Matchers.hasSize(expectedBrokerCount));
+        var knownReady = Collections.synchronizedSet(new HashSet<String>());
+        var toProbe = Collections.synchronizedSet(new HashSet<String>());
+
+        var originalBootstrap = String.valueOf(connectionConfig.get(BOOTSTRAP_SERVERS_CONFIG));
+        toProbe.addAll(Arrays.asList(originalBootstrap.split(",")));
+
+        while(knownReady.size() < expectedBrokerCount) {
+            if (toProbe.isEmpty()) {
+                throw new IllegalArgumentException(String.format("Ran out of addresses to probe before cluster ready: %s", originalBootstrap));
+            }
+            var probeAddress = toProbe.iterator().next();
+
+            var copy = new HashMap<>(connectionConfig);
+            copy.put(BOOTSTRAP_SERVERS_CONFIG, probeAddress);
+
+            try (Admin admin = Admin.create(copy)) {
```

Comment:
Feels odd to be re-creating the admin client every iteration.

I haven't checked what happens if it fails to connect in one iteration what happens on a subsequent one.

Reply:
> Feels odd to be re-creating the admin client every iteration.
I suspect that we could move the admin client outside the retry loop.  I don't feel an imperative to change tbh

Reply:
Each `create` causes a [new thread](https://github.com/apache/kafka/blob/5c52c61a46f103cda7fa48bd96584dafe984d6d3/clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java#L596) to be created which is less than ideal.